flopy/mf6/utils/codegen/renderable.py

```diff
@@ -0,0 +1,151 @@
+"""
+This module contains a decorator intended to
+allow modifying dataclass instances to make
+them more palatable for templates. It also
+keeps implementation details incidental to
+the current design of MF6 input framework
+cleanly isolated from the reimplementation
+of which this code is a part, which aims
+for a more general approach.
+
+Jinja supports attribute- and dictionary-
+based access on arbitrary objects but does
+not support arbitrary expressions, and has
+only a limited set of custom filters; this
```

Comment:
I was able to get rid of another piece of the shim, I think we can use more of jinja's functionality.

First I made a python file flopy/mf6/utils/codegen/custom_tests.py.

```py
from jinja2 import pass_context


@pass_context
def init_model_variable(ctx, key) -> bool:
    skip = ["packages", "export_netcdf", "nc_filerecord"]
    refs = ctx.get("meta", dict()).get("refs", dict())
    if any(refs) and ctx["name"] != (None, "nam"):
        for k in refs.keys():
            if ctx["vars"].get(k, None):
                skip.append(k)
    return key not in skip


@pass_context
def init_package_variable(ctx, key) -> bool:
    name = ctx["name"]
    skip = []
    if name.r == "nam":
        skip = ["export_netcdf", "nc_filerecord"]
    elif name == ("utl", "ts"):
        skip = ["method", "interpolation_method_single", "sfac"]
    return key not in skip


def init_simulation_variable(key) -> bool:
    skip = [
        "tdis6",
        "models",
        "exchanges",
        "mxiter",
        "solutiongroup",
    ]
    return key not in skip
```

These separate functions can be used in the Environment.tests:

```py
    loader = __jinja.PackageLoader("flopy", "mf6/utils/codegen/templates/")
    env = __jinja.Environment(loader=loader)
    env.tests["init_simulation_variable"] = (
        custom_tests.init_simulation_variable
    )
    env.tests["init_model_variable"] = custom_tests.init_model_variable
    env.tests["init_package_variable"] = custom_tests.init_package_variable
```

Then you can use it directly in the jinja file like so:

```jinja
{%- for n, var in vars.items() if n is init_model_variable %}
```

instead of the old: `if n not in init_skip`.

And you can get rid of the `_init_skip` function in the shim.

Reply:
I think with filters/tests we can get rid of the shim entirely. Nice..

Reply:
dropped the shim in favor of filters and tests. I should have RTFM!